trin-core/src/types/receipts.rs

```diff
@@ -0,0 +1,523 @@
+    }
+}
+
+#[derive(Eq, Hash, Debug, Copy, Clone, PartialEq)]
+#[repr(u8)]
+pub enum TypedTxId {
+    EIP1559Transaction = 0x02,
+    AccessList = 0x01,
+    Legacy = 0x00,
+}
+
+impl TypedTxId {
+    pub fn try_from_wire_byte(n: u8) -> anyhow::Result<Self> {
+        match n {
+            x if x == TypedTxId::EIP1559Transaction as u8 => Ok(Self::EIP1559Transaction),
+            x if x == TypedTxId::AccessList as u8 => Ok(Self::AccessList),
+            x if (x & 0x80) != 0x00 => Ok(Self::Legacy),
+            _ => Err(anyhow!("Invalid byte selector for tx type.")),
+        }
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum TypedReceipt {
+    Legacy(LegacyReceipt),
+    AccessList(LegacyReceipt),
+    EIP1559Transaction(LegacyReceipt),
+}
+
+impl TypedReceipt {
+    /// Create a new receipt.
+    pub fn new(type_id: TypedTxId, legacy_receipt: LegacyReceipt) -> Self {
+        //curently we are using same receipt for both legacy and typed transaction
+        match type_id {
+            TypedTxId::EIP1559Transaction => Self::EIP1559Transaction(legacy_receipt),
+            TypedTxId::AccessList => Self::AccessList(legacy_receipt),
+            TypedTxId::Legacy => Self::Legacy(legacy_receipt),
+        }
+    }
+
+    pub fn tx_type(&self) -> TypedTxId {
+        match self {
+            Self::Legacy(_) => TypedTxId::Legacy,
+            Self::AccessList(_) => TypedTxId::AccessList,
+            Self::EIP1559Transaction(_) => TypedTxId::EIP1559Transaction,
+        }
+    }
+
+    pub fn receipt(&self) -> &LegacyReceipt {
+        match self {
+            Self::Legacy(receipt) => receipt,
+            Self::AccessList(receipt) => receipt,
+            Self::EIP1559Transaction(receipt) => receipt,
+        }
+    }
+
+    pub fn receipt_mut(&mut self) -> &mut LegacyReceipt {
+        match self {
+            Self::Legacy(receipt) => receipt,
+            Self::AccessList(receipt) => receipt,
+            Self::EIP1559Transaction(receipt) => receipt,
+        }
+    }
+
+    fn decode(tx: &[u8]) -> Result<Self, DecoderError> {
+        if tx.is_empty() {
+            // at least one byte needs to be present
+            return Err(DecoderError::RlpIncorrectListLen);
+        }
+        let id = TypedTxId::try_from_wire_byte(tx[0]);
+        if id.is_err() {
+            return Err(DecoderError::Custom("Unknown transaction"));
+        }
+        //other transaction types
+        match id.unwrap() {
+            TypedTxId::EIP1559Transaction => {
+                let rlp = Rlp::new(&tx[1..]);
+                Ok(Self::EIP1559Transaction(LegacyReceipt::decode(&rlp)?))
+            }
+            TypedTxId::AccessList => {
+                let rlp = Rlp::new(&tx[1..]);
+                Ok(Self::AccessList(LegacyReceipt::decode(&rlp)?))
+            }
+            TypedTxId::Legacy => Ok(Self::Legacy(LegacyReceipt::decode(&Rlp::new(tx))?)),
+        }
+    }
+
+    pub fn decode_rlp(rlp: &Rlp) -> Result<Self, DecoderError> {
```

Comment:
Same comment as above for `LegacyReceipt`. Why not a separate `impl` block for the `rlp` traits?

Reply:
I've implemented them for `LegacyReceipt`, but turns out it's not so easy for `Receipt`. The problem is with encoding. The encoding api...
```rust
fn rlp_append(&self, s: &mut RlpStream);
```
does not play nicely with prefixing the typed transaction id to the rlp encoded transaction. I've tried using some of the available methods on `RlpStream` to make it work. Some methods like `append_raw()` and `append_internal()` seemed promising, but I had no luck. If you add an item to `RlpStream`, it always treats it as an encoded item, and won't simply allow a byte prefix along with an rlp-encoded hex string. Obvs, it's possible I'm missing something... But for now it doesn't seem worth spending more time getting those traits supported (even our `Header` implementation doesn't use `Decodable`). Implementing `Decodable` and not `Encodable` for `Receipt` seems a bit wonky to me, mixing up apis. So for `Receipt` I'd prefer to just leave the `Receipt::encode` and `Receipt::decode` implementations.